lineax/_operator.py

```diff
@@ -906,6 +906,78 @@ def out_structure(self):
         return self.operator.out_structure()
 
 
+class WoodburyLinearOperator(AbstractLinearOperator, strict=True):
+    """As [`lineax.MatrixLinearOperator`][], but for specifically a matrix
+    with A + U C V structure, such that the Woodbury identity can be used"""
+
+    A: AbstractLinearOperator
+    C: Inexact[Array, " k k"]
+    U: Inexact[Array, " n k"]
+    V: Inexact[Array, " k n"]
+    UCV: Array
+    tags: frozenset[object] = eqx.field(static=True)
+
+    def __init__(
+        self,
+        A: AbstractLinearOperator,
+        C: Inexact[Array, " k k"],
+        U: Inexact[Array, " n k"],
+        V: Inexact[Array, " k n"],
+        tags: Union[object, frozenset[object]] = (),
+    ):
+        """**Arguments:**
+
+        Matrix of form A + U C V, such that the inverse can be computed
+        using Woodbury matrix identity
+
+        - `A`: Linear operator, in/out shape (n,n)
+        - `C`: A rank-two JAX array. Shape (k,k)
+        - `U`: A rank-two JAX array. Shape (n,k)
+        - `V`: A rank-two JAX array. Shape (k,n)
+
+        """
+        self.A = A
+        self.C = inexact_asarray(C)
+        self.U = inexact_asarray(U)
+        self.V = inexact_asarray(V)
+        (N, M) = self.A.in_structure(), self.A.out_structure()
+        if N != M:
```

Comment:
Edge-case: the structures are arbitrary PyTrees, and we don't know whether they might have defined an `__eq__` method that does something else.

Instead we should use `eqx.tree_equal` here.